keps/sig-scheduling/3838-pod-mutable-scheduling-directives/README.md

```diff
@@ -0,0 +1,969 @@
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Allow mutating node affinity/selector of pods unconditionally.
+- Restrict updating tolerations on scheduling gates. We already allow adding tolerations 
+  unconditionally, and so restricting that to gated pods is not a backward compatible change.
+- Allow mutating pod (anti)affinity constraints. It is not clear how this can be done without
+  potentially violating a policy check that was previously done at pod CREATE.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Node affinity and selector are currently immutable for Pods. The proposal is to relax this validation
+for Pods that are blocked on a scheduling readiness gate (i.e., not yet considered for scheduling by 
+kube-scheduler). Note that adding tolerations and mutating labels and annotations are
+already allowed.
+
+This has no impact on applications controllers (like the ReplicaSet controller). Application controllers
+generally has no dependency on the scheduling directives expressed in the application's pod template, and 
+they don't reconcile existing pods if they changed.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+I want to build a controller that implements workload queueing and influences when and where a 
+workload should run. The workload controller creates the pods of a workload instance. To control 
+when and where the workload should run, I have a webhook that injects a scheduling readiness gate, 
+and a controller that removes the gate when capacity becomes available.
+
+At the time of creating the workload, it may not be known on which subset of the nodes (e.g., a zone or 
+a VM type) a workload should run. To control where a workload should run, the controller updates 
+the node selector/affinity of the workload's pods; by doing so, the queue controller is able to influence the 
+scheduler and autoscaler decisions.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
Another risk (or note): fields are atomic, meaning that they can't be added via SSA independently and the controller owns the entire nodeSelector field.

Reply:
If two controllers try to update at the same time, one of them will get a conflict and can try again, and this is not different to other mutable atomic fields, so I am not sure why this is worth noting here?

Reply:
if the two controllers use SSA, they might both succeed and overwrite each others' update. Same with Job template actually.

Did I get that right @liggitt ?

Reply:
there's two issues... 

1. whether the fields being updated support multiple SSA owners - atomic fields don't
2. how the controllers are updating, and whether they're trying to avoid conflicts - controllers have multiple ways they can update:
  * SSA with ?force=false - this would return a conflict error if they touched an atomic field owned by someone else
  * SSA with ?force=true - this would stomp the atomic field owned by someone else with what they applied and steal ownership
  * read/update/write - this would stomp the field with what they wrote and steal ownership 
  * other patch types (jsonpatch, mergepatch, etc) - this would patch the field with what they wrote and steal ownership

Reply:
If those semantics are documented somewhere, then this KEP can refer to it, but I don't think it should document them here since it isn't proposing something that changes them.

Reply:
Yeah, I agree

Reply:
https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy

What is relevant in this KEP is that the fields are atomic.